hpat/hiframes_typed.py

```diff
@@ -751,6 +754,96 @@ def f(arr, on_arr, w, center):  # pragma: no cover
             return nodes
         return [assign]
 
+    def _handle_series_combine(self, assign, lhs, rhs, series_var):
+        """translate s1.combine(s2,lambda x1,x2 :...) to prange()
+        """
+        func = guard(get_definition, self.func_ir, rhs.args[1])
+        if func is None or not (isinstance(func, ir.Expr)
+                                and func.op == 'make_function'):
+            raise ValueError("lambda for combine not found")
+
+        out_typ = self.typemap[lhs.name].dtype
+
+        # If we are called with 3 arguments, we must use 3rd arg as a fill value,
+        # instead of Nan.
+        use_nan = len(rhs.args) == 2
+
+        # prange func to inline
+        if use_nan:
+            func_text = "def f(A, B):\n"
+        else:
+            func_text = "def f(A, B, C):\n"
+        func_text += "  numba.parfor.init_prange()\n"
+        func_text += "  n1 = len(A)\n"
+        func_text += "  n2 = len(B)\n"
+        func_text += "  n = n1\n"
+        func_text += "  if n2 > n:\n"
+        func_text += "      n = n2\n"
+        func_text += "  S = numba.unsafe.ndarray.empty_inferred((n,))\n"
+        func_text += "  for i in numba.parfor.internal_prange(n):\n"
+        if use_nan:
+            func_text += "    t1 = np.nan\n"
```

Comment:
`np.nan` cannot be stored in integer arrays. I think we could add a runtime assertion for integer output cases that prevents different series sizes (with a message asking for explicit float type annotation). This assertion can be after `init_prange` not to prevent optimizations.

Reply:
Only the dtype of the output matters for this NaN handling. The output can be float with even both inputs being integer (e.g. a formula with division).

Reply:
Pandas doesn't check type of inputs for conversion: https://github.com/pandas-dev/pandas/blob/master/pandas/core/series.py#L2303

Reply:
In that case, I'm not sure why we should check series length in combine handler at all. map_func can return valid integer even if one of the inputs is Nan.

Reply:
When sizes don't match, NaN values need to be assigned. Pandas throws an error when it can't assign NaN to integer output array:
```python
S1 = pd.Series([1,2,3,4])
S2 = pd.Series([1,2,3,4,5])
S1.combine(S2, lambda a,b: a+b)
```
`ValueError: cannot convert float NaN to integer`

Reply:
I understand that, but map_func can return integer for Nan input, as a trivial example consider:
```
S1 = pd.Series([1,2,3,4])
S2 = pd.Series([1,2,3,4,5])
S1.combine(S2, lambda a,b: b)  # works fine with pandas
```

Reply:
Ok, I see what you mean, it's not necessarily related to length. Sounds like a much harder check to always examine the output of the user function, which could have performance implications. I think we can skip the check. Thanks.

Reply:
After thinking about it some more, we probably need an assert, because otherwise inputs of map_func are not type stable if input series is not float (input has either float type due to nan or input series type). I've tried removing assertions and as expected it broke integer combine testes.

Reply:
Ok. makes sense.